ipapython/ipautil.py

```diff
@@ -282,6 +282,37 @@ def write_tmp_file(txt):
 
     return fd
 
+
+@contextmanager
+def concurrent_open(path, mode='r'):
+    """Ensure that complete file is read/written
+
+    Works only for r, rb, w, w+, wb, wb+ modes. Default is r.
+
+    In read mode behaves the same as build-in open. In write mode all data are
+    written to the temporary file first and then moved to target path.
+    This approach ensures that reading will be performed on complete file and
+    writting will result in complete or none file written at all.
+    """
+    if mode in ('w', 'w+', 'wb', 'wb+'):
+        with tempfile.NamedTemporaryFile(
+                mode=mode, prefix=path, delete=False) as temp_file:
```

Comment:
Ah, I didn't see that you wrapped ```concurrent_open``` in a ```@contextmanager```. In that case the ```__exit__``` function of ```NamedTemporaryFile``` is not called because the function doesn't return the ```NamedTemporaryFile``` instance directly but rather a wrapped instance.

try/except OSError is fine. You don't want to call ```os.unlink()``` when ```os.rename()``` was successfully called.